### PR TITLE
Projectsetup

### PR DESCRIPTION
**Pull request by James-Large**
In draft. Bold indicates main open questions seeking review

#### Reference Issues/PRs
#16, #21  To be closed after this is agreed upon

#### What does this implement/fix? Explain your changes.
General update of project structure, testing procedures etc. 

Main things: 

- Migrated "import keras" calls over to "from tensorflow import keras" #21 
  - Updated some related API calls, e.g. fcn/_base.py : 'keras.layers.pooling.GlobalAveragePooling1D'    to just    'keras.layers.GlobalAveragePooling1D'
- Updated the setup.py for these changes and newer versions/compatabilities #16
  - tf.keras changes move up the minimum required tf version from 1.8 to 1.9, setup and readme reflects this
  - tf 1.13+ is compatible with python 3.7 (AND 3.6), have updated setup.py to reflect this, **but haven't done anything intelligent with it (e.g. if tf < 1.13 already installed, enforce 3.6 but not 3.7)**
  - tensorflow-addons is the replacement for keras-contrib for tf >=2.1. It is installed if installing the latest version of tensorflow, otherwise keras-contrib is still needed. Installation instructions explain this. Only place currently in code this affects things is in encoder https://github.com/sktime/sktime-dl/compare/dev...projectsetup?expand=1#diff-8ceef7048a114b8be1f0d4885c2cf45eR3 
  - **Commented out the requirement for numpy to be installed prior to sktime-dl. Do we still need this?** Only usage was for the include_dirs setup field. Removing this requirement (and the include_dirs arg) PRIOR to setup.py and installing it normally THROUGH setup.py seems to work. 
  - Updated my local envs and have tested sktime-dl for tf 1.9, tf 1.15, tf 2.1
- Updated travis tests
  - Edited the matrices, now have ones for tf 1.9 (minimum supported version), 1.15 (last 1.x) and 2.1 (most recent)
  - Changed method and location of keras-contrib installation during testing (needed for tf<2.1). Unsure whether this will work, will find out in the pr tests. **May need/want to do something more intelligent with this, including not installing when on the tf2.1 env**
- Updated readme installation instructions and overview

Notes: 
- Ran the test_accuracy tests a lot during these changes, 3 out of maybe 25 times the FCNClassifier or TWIESNClassifier only just failed  (e.g. .FCNClassifier accuracy: 0.9543245869776482,   lower=0.961-2*0.003). Unsure what to do with this. 2 std devs suggests a test may fail 5% of the time in any case. I like the within-stddev nature of these tests, but there is still the possibility of a 'scare' happening due to chance. **Wonder about enforcing within 3 or 4 times the stddev. Thoughts?** 
- Have added a warning statement in setup.py about installing keras-contrib if tf<2.1 is already installed, but the message (nor any print/warning statement) does not seem to appear in the prompt. Maybe investigate, or just drop it, the install instructions are on the readme

### CODE REVIEW

**Comment by Withington** [comment]
That's a neat way to switch from keras_contrib to tensorflow_addons.

This PR didn't trigger a build and it won't build on my fork either. It could be a Travis issue - not picking up new branches at this time.

Will this work with TF 2.0? I got errors locally on ```test_encoder_accuracy```.

On "_Wonder about enforcing within 3 or 4 times the stddev. Thoughts?_" - 3x std dev is reasonable. Perhaps the number of flaky reruns should then be reduced from 3 to 2?

File: .travis.yml [hunk 1]
```diff
@@ -37,14 +38,22 @@ install:
...
+    # pip install git+https://github.com/keras-team/keras-contrib.git@master
+    
+    # now need to install keras-contrib for tf.keras instead of standalone keras
+    # not needed for the tf_version 2.1 env, but does not hurt either. investigate 
+    # conditional installation 
+    - git clone https://www.github.com/keras-team/keras-contrib.git
+    - cd keras-contrib
+    - python convert_to_tf_keras.py
+    - USE_TF_KERAS=1 python setup.py install
+
```
Comment (Withington) [inline .travis.yml]: This tab is what is preventing Travis from running. It needs to be deleted. 
I don't have write permission to do that.

**Comment by James-Large** [comment]
Good catch on the tab, had to put this down after pushing so only looking at it now 

All three envs failing annoyingly, all for different reasons:

- tf1.9 just needs h5py, which I didn't notice isn't on the 1.9 dependency tree I guess, is on the other 2. Will just add it to requires explicitly
- tf1.15 cannot figure out if we're using keras or tf.keras seemingly. Maybe tf-addons works with 1.15 (haven't seen anything about that though), will investigate
- tf2.1 is weird. A copy of the error (bold added)

________ ERROR collecting sktime_dl/deeplearning/tests/test_accuracy.py ________
ImportError while importing test module '/home/travis/build/sktime/sktime-dl/sktime_dl/deeplearning/tests/test_accuracy.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../../miniconda/envs/testenv/lib/python3.7/site-packages/sklearn/__check_build/__init__.py:44: in <module>
    from ._check_build import check_build  # noqa
E   ImportError: dlopen: cannot load any more object with static TLS
During handling of the above exception, another exception occurred:
../sktime_dl/deeplearning/__init__.py:1: in <module>
    from .cnn._classifier import CNNClassifier
../sktime_dl/deeplearning/cnn/_classifier.py:5: in <module>
    from sktime_dl.deeplearning.base.estimators import BaseDeepClassifier
../sktime_dl/deeplearning/base/estimators/__init__.py:1: in <module>
    from ._classifier import BaseDeepClassifier
../sktime_dl/deeplearning/base/estimators/_classifier.py:24: in <module>
    from sktime.classifiers.base import BaseClassifier
../../../../miniconda/envs/testenv/lib/python3.7/site-packages/sktime/classifiers/base.py:1: in <module>
    from sklearn.base import BaseEstimator
../../../../miniconda/envs/testenv/lib/python3.7/site-packages/sklearn/__init__.py:81: in <module>
    from . import __check_build  # noqa: F401
../../../../miniconda/envs/testenv/lib/python3.7/site-packages/sklearn/__check_build/__init__.py:46: in <module>
    raise_build_error(e)
../../../../miniconda/envs/testenv/lib/python3.7/site-packages/sklearn/__check_build/__init__.py:41: in raise_build_error
    %s""" % (e, local_dir, ''.join(dir_content).strip(), msg))
E   ImportError: dlopen: cannot load any more object with static TLS
E   ___________________________________________________________________________
E   Contents of /home/travis/miniconda/envs/testenv/lib/python3.7/site-packages/sklearn/__check_build:
E   setup.py                  __init__.py               __pycache__
E   _check_build.cpython-37m-x86_64-linux-gnu.so
E   ___________________________________________________________________________
E   **It seems that scikit-learn has not been built correctly.
E   
E   If you have installed scikit-learn from source, please do not forget
E   to build the package before using it: run `python setup.py install` or
E   `make` in the source directory.
E   
E   If you have used an installer, please check that it is suited for your
E   Python version, your operating system and your platform.**


Encountered anything like this with normal sktime @mloning ? Even if so, sklearn is being installed via sktime so if you had a fix it should be being applied. Typical early search results just say downgrade sklearn as usual.  Didn't see anything like that on my windows box.

**Comment by mloning** [comment]
Happy to merge this now. I think we can add CI for Windows and macOS some time later.

**Comment by James-Large** [comment]
Cool. Yeah unless someone else wants to bring it in sooner, I'm leaving the alternate os testing for later. It would not be trivial, at least. Also, I did a quick test with python 3.8, don't think that's going to work with current tf versions 

I think we're ready for a master push within another days or two's work, whenever those can happen. I need to do a bit of usability updating with the ensemble class for someone here, and in general we should add some basic usage examples at this point. 

Questionable for 0.2.0 or for afterwards is bringing back in 'proper' (however we define that) support for callbacks and how that would work

Issues for these and other things will be made and I'll copy some of this comment over to dev pr too

**Comment by James-Large** [comment]
Squashing/merging, and will close #16 and #21

**Pull request merged**